assets/app/view/game/part/track_node_path.rb

```diff
@@ -0,0 +1,150 @@
+# frozen_string_literal: true
+
+require 'view/game/part/base'
+
+module View
+  module Game
+    module Part
+      class TrackNodePath < Base
+        needs :tile
+        needs :path
+        needs :color, default: 'black'
+        needs :width, default: 8
+        needs :dash, default: '0'
+
+        EDGE_PERP_ANGLES = [90, 30, -30, -90, -150, 150].freeze
+
+        EDGE_REGIONS = {
+          0 => TRACK_TO_EDGE_0,
+          1 => TRACK_TO_EDGE_1,
+          2 => TRACK_TO_EDGE_2,
+          3 => TRACK_TO_EDGE_3,
+          4 => TRACK_TO_EDGE_4,
+          5 => TRACK_TO_EDGE_5,
+        }.freeze
+
+        def load_from_tile
+          # for now assumes one edge and one node on path
+          @edge = @path.edges.first.num
+          @terminal = @path.terminal
+          @junction = @path.junction
+          return unless @path.stop
+
+          @stop = @path.stop
+          @ct_edge = @tile.preferred_city_town_edges[@stop]
+        end
+
+        def begin_location
+          {
+            x: -Math.sin((@edge * 60) / 180 * Math::PI) * 87,
+            y: Math.cos((@edge * 60) / 180 * Math::PI) * 87,
+          }
+        end
+
+        def end_location
+          return { x: 0, y: 0 } if @junction
+
+          if (@tile.towns.size + @tile.cities.size) < 2
+            # this really needs to be stored in the city/town object
+            full_distance = 0
+          else
+            full_distance = 50
+            full_distance -= 15 if @tile.borders.any? { |border| border.edge == @ct_edge }
+          end
+          {
+            x: -Math.sin((@ct_edge * 60) / 180 * Math::PI) * full_distance,
+            y: Math.cos((@ct_edge * 60) / 180 * Math::PI) * full_distance,
+          }
+        end
+
+        def control_location
```

Comment:
I'm wondering if it would be possible to memoize `start_location`, `end_location`, and `control_location` somewhere to save some recalculations. I added a few quick debug lines, and `control_location` had a good number of repeated calls where `begin_location`, `end_location`, and `@edge` all had the same values.

Haven't thought about it too much, they'd probably need to be pulled out of `View::Game::Part` and put higher up in the module hierarchy.

Reply:
@tobymao you have any thoughts on that?

Reply:
I wasn't sure how to do this. Perhaps storing in the path object?

Reply:
begin_location could definitely be a constant of edge => map

end_location looks more dynamic, you can memoize doing 

def end_location
  @end_location ||= begin 
...
end

you can grep for other examples

Reply:
it would be easier to set all these up in the setup method as well

Reply:
load_from_tile

Reply:
So my longer term plan is for begin_location to resemble end_location more since eventually it will have to handle paths that don't go to edges (city/town to city/town paths on a hex). I'll change to have these cache in addition to control_location.